lib/upipe-modules/upipe_auto_inner.c

```diff
@@ -0,0 +1,388 @@
+                upipe_warn(upipe, "fail to allocate inner pipe");
+                continue;
+            }
+        }
+        int ret = upipe_set_flow_def(inner, flow_def);
+        if (unlikely(!ubase_check(ret))) {
+            upipe_dbg(inner, "flow def rejected, trying next manager");
+            upipe_release(inner);
+            inner = NULL;
+            continue;
+        }
+        else {
+            upipe_dbg(inner, "flow def accepted");
+        }
+        break;
+    }
+
+    if (unlikely(!inner))
+        return UBASE_ERR_INVALID;
+
+    upipe_autoin_store_bin_output(upipe, inner);
+    upipe_autoin_store_bin_input(upipe, upipe_use(inner));
+    return UBASE_ERR_NONE;
+}
+
+/** @internal @This handles control commands.
+ *
+ * @param upipe description structure of the pipe
+ * @param command control command to handle
+ * @param args optional arguments
+ * @return an error code
+ */
+static int upipe_autoin_control(struct upipe *upipe, int command, va_list args)
+{
+    switch (command) {
+        case UPIPE_SET_FLOW_DEF: {
+            struct uref *flow_def = va_arg(args, struct uref *);
+            return upipe_autoin_set_flow_def(upipe, flow_def);
+        }
+
+        case UPIPE_REGISTER_REQUEST:
+        case UPIPE_UNREGISTER_REQUEST:
+        case UPIPE_BIN_GET_FIRST_INNER:
+            return upipe_autoin_control_bin_input(upipe, command, args);
+        case UPIPE_GET_OUTPUT:
+        case UPIPE_SET_OUTPUT:
+        case UPIPE_BIN_GET_LAST_INNER:
+            return upipe_autoin_control_bin_output(upipe, command, args);
+    }
+    return upipe_autoin_control_inner(upipe, command, args);
+}
+
+/** @internal @This finds an inner pipe manager item already added.
+ *
+ * @param mgr auto inner pipe manager
+ * @param inner_mgr inner pipe manager to find
+ * @return the inner manager item if found, NULL otherwise
+ */
+static struct upipe_autoin_mgr_item *
+upipe_autoin_mgr_find(struct upipe_mgr *mgr, struct upipe_mgr *inner_mgr)
+{
+    struct upipe_autoin_mgr *upipe_autoin_mgr = upipe_autoin_mgr_from_mgr(mgr);
+
+    struct uchain *uchain;
+    ulist_foreach(&upipe_autoin_mgr->inner_mgrs, uchain) {
+        struct upipe_autoin_mgr_item *item =
+            upipe_autoin_mgr_item_from_uchain(uchain);
+        if (item->mgr == inner_mgr)
+            return item;
+    }
+    return NULL;
+}
+
+/** @internal @This adds a pipe manager to the auto inner pipe manager.
+ *
+ * @param mgr auto inner pipe manager
+ * @param name name use for inner uprobe prefix
+ * @param inner_mgr pipe manager to attach
+ * @return an error code
+ */
+static int _upipe_autoin_mgr_add_mgr(struct upipe_mgr *mgr,
+                                     const char *name,
+                                     struct upipe_mgr *inner_mgr)
+{
+    struct upipe_autoin_mgr *upipe_autoin_mgr = upipe_autoin_mgr_from_mgr(mgr);
+
+    if (unlikely(inner_mgr == NULL))
+        return UBASE_ERR_INVALID;
+
+    if (unlikely(upipe_autoin_mgr_find(mgr, inner_mgr)))
```

Comment:
You should only allow this if urefcount_single(&mgr->refcount). Otherwise there is a race condition with already allocated pipes. See upipe_filter_format.